rfcs/2022-05-17-11532-chronicle-sink.md

```diff
@@ -0,0 +1,215 @@
+  unstructured log events.
+
+### Out of scope
+
+- Adding any explicit knowledge of SIEM events as a separate event
+  type into Vector will not be considered.
+
+## Pain
+
+## Proposal
+
+### User Experience
+
+This adds two new sinks to Vector to send data into Google Chronicle.
+
+#### UDM (Unified Data Model) events
+
+The `gcp_chronicle_udm` sink.
+
+https://cloud.google.com/chronicle/docs/reference/ingestion-api#udmevents
+https://cloud.google.com/chronicle/docs/unified-data-model/udm-usage
+
+UDM events are events that conform to a [strict schema](https://cloud.google.com/chronicle/docs/reference/udm-field-list).
+An example event:
+
+```json
+{
+  "metadata": {
+    "event_timestamp": "2019-10-23T04:00:00.000Z",
+    "event_type": "NETWORK_HTTP",
+    "product_name": "Acme Proxy",
+    "vendor_name": "Acme"
+  },
+  "network": {
+    "http": {
+      "method": "GET",
+      "response_code": 200
+    }
+  },
+  "principal": {
+    "hostname": "host0",
+    "ip": "10.1.2.3",
+    "port": 60000
+  },
+  "target": {
+    "hostname": "www.altostrat.com",
+    "ip": "198.51.100.68",
+    "port": 443,
+    "url": "www.altostrat.com/images/logo.png"
+  }
+}
+```
+
+Events can be batched and sent in a JSON object that also includes the
+`customer_id`:
+
+```json
+{
+"customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+"events": [...]
+}
+```
+
+UDM events can be sumbitted as either JSON or Proto3. We will submit using JSON
+as this fits in easier with Vectors data model. The newly developed schema
+support for Vector will be used to ensure the JSON is formed according to the
+[Chronicle requirements](https://cloud.google.com/chronicle/docs/unified-data-model/format-events-as-udm).
+
+Vector has no internal knowledge of SIEM events so it will be up to the user to
+ensure the event is well formed - either the source will receive the data in the
+correct structure, or it will be transformed into such before reaching the sink.
+
+#### Unstructured log data events
+
+The `gcp_chronicle_unstructured` sink.
+
+Unstructured log events are events in plain text with no strict schema attached.
+
+```json
+{
+  "log_text": "26-Feb-2019 13:35:02.187 client 10.120.20.32#4238: query: altostrat.com IN A + (203.0.113.102)",
+  "ts_epoch_microseconds": 1551188102187000
+}
+```
+
+Events are batched and sent in a JSON object that includes the `customer_id` and
+a `log_type`:
+
+```json
+{
+  "customer_id": "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c",
+  "log_type": "BIND_DNS",
+  "entries": [...]
+}
+```
+
+`log_type` has to be one of the values returned by the [logtypes endpoint](https://cloud.google.com/chronicle/docs/reference/ingestion-api#logtypes).
+
+
+If `event_type == "unstructured"` we will require another configuration field -
+`log_type` to be set. This field will be templatable. The field can be removed
+from the final event via `encoding.exclude_fields`.
+
+The encoding of the message (text or json) can be set via an `encoding` field.
```

Comment:
Given the way the docs read for unstructured events, I'm not sure I see a reason to expose the ability to change the encoding of the event?

Conceivably the whole point is to take a raw message (i.e. `.message`) and just pass it through, as it was emitted from the upstream device/client/etc. If we allow control of the encoding, now we might be adding extra fields that change the output.

I'd say that we should not expose the encoding of the sink, and if users _really_ want to do that, they should use a remap transform.

Reply:
Looking at the [parsers supported by Chronicle](https://cloud.google.com/chronicle/docs/ingestion/parser-list/supported-default-parsers) a large number of them are ultimately Json formats. So it does seem like supporting Json encoding could be quite useful.  

Arguably, we could extend the codecs to the other formats - CSV etc.. but I think that is out of scope for the initial implementation. If user demand is there it would be worth looking into.